Fixes #32750 - make searchbar css looks like PF4

Ron-Lavi opened the pull request:
As the first step of moving it to PF4,
we can already override it with CSS and apply some changes to pages,

React pages are the only ones that would be affected at the moment,
legacy pages will still have the regular bootstrap3 design.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

Ron-Lavi commented on app/helpers/search_bar_helper.rb:
```diff
@@ -13,6 +13,7 @@ def mount_search_bar
                         searchQuery: params[:search],
                         url: url,
                         useKeyShortcuts: true,
+                        isPF4: false,
```
should the legacy pages also use the new look or have the PF3 bootstrap design?

MariaAga replied:
Can you add screenshots on how it looks? :)

Ron-Lavi replied:
If it will be the default this is how it would look:

![gnome-shell-screenshot-928P80](https://user-images.githubusercontent.com/26363699/131486386-58b4dfbf-1c87-4ba2-a2ea-8903196a1235.png)

maybe it will give us a push with replacing buttons all over :)

tbrisker replied:
Maybe I missed something, why do we want the legacy pages to have a different style? IMHO one of the biggest strengths of the ruby helpers is that we can change everywhere at once and plugins don't need to care.

Ron-Lavi replied:
because on old .erb pages the table and buttons on the page are still in bootstrap3 style,
I think it will just not look well in most of the pages until we switch those components to PF4 style,
when we'll do that, we can just tweak the `isPF4` prop.

tbrisker replied:
@MariSvirik what do you think? should we have different looking search fields on different pages, or consistent search fields everywhere, but with styling that is different from the rest of the page on legacy pages?
Personally, I tend to prefer the second, and gradually transform legacy pages, but not sure what is correct from UX standpoint. That approach also simplifies the code quite a bit.

MariSvirik replied:
@tbrisker Based on the discussion with other designers we decided to go with this option:  'having the new component with pf4 styling on all pages, including legacy ones - this mean the search buttons will not match the action buttons until the buttons are updated to pf4 as well on those pages, but the search component will be the same everywhere' 

This would put more pressure on updating the legacy pages later on.

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/Audits/AuditsPage/AuditsPage.js:
```diff
@@ -26,12 +26,13 @@ const AuditsPage = ({
+        target="_blank"
+        variant="secondary"
       >
-        <Icon type="pf" name="help" />
```
No need for icons in buttons, right? @marisvirik

MariSvirik replied:
which buttons are you referring to, @laviro ? But not really.

Ron-Lavi replied:
Before:
![gnome-shell-screenshot-OHVG80](https://user-images.githubusercontent.com/26363699/131486773-0bdf75de-3faa-4c65-8861-60c1a9f88b2b.png)

After:
![gnome-shell-screenshot-XBAJ80](https://user-images.githubusercontent.com/26363699/131486867-676cd6c1-affc-46bc-b327-180a37edce07.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/Audits/AuditsPage/AuditsPage.js:
```diff
@@ -26,12 +26,13 @@ const AuditsPage = ({
         rel="external noreferrer noopener"
-        className="btn-docs"
+        target="_blank"
+        variant="secondary"
```
![Screenshot](https://user-images.githubusercontent.com/26363699/130983651-92d4f1b1-1b59-439d-9038-ff8dd248577d.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/Models/ModelsPage/ModelsPage.js:
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { Button } from 'patternfly-react';
+import { Button } from '@patternfly/react-core';
```
![Screenshot (1)](https://user-images.githubusercontent.com/26363699/130983824-b00101bb-b8b5-4ce6-b259-b2ac9a563698.png)

cc @xprazak2

MariSvirik replied:
use sentence case in a table header: 'Vendor class', 'Hardware model'
In buttons as well: 'Create model'

Ron-Lavi replied:
Thanks, fixed in the second commit: https://github.com/theforeman/foreman/pull/8743/commits/b45316e9564dea889eb01eff841929d256a004ea

tbrisker replied:
That doesn't seem related to this PR, please extract to a separate one (which would be a trivial merge).

Ron-Lavi replied:
sure, no problems.

MariaAga commented:
@laviro Can you add screenshots that show pf3 vs 4?

Ron-Lavi commented:
> @laviro Can you add screenshots that show pf3 vs 4?

Audits page:
- before:
![gnome-shell-screenshot-A8NW80](https://user-images.githubusercontent.com/26363699/131500113-b17551c2-412e-498b-b68a-5ed4596da908.png)

- after:
![](https://user-images.githubusercontent.com/26363699/130983651-92d4f1b1-1b59-439d-9038-ff8dd248577d.png)

Models page:
- before:
![gnome-shell-screenshot-O0RZ80](https://user-images.githubusercontent.com/26363699/131500295-6d325f5f-cc77-4a67-ab5f-e65de828577d.png)

- after:
![](https://user-images.githubusercontent.com/26363699/130983824-b00101bb-b8b5-4ce6-b259-b2ac9a563698.png)

MariaAga commented:
Waiting for a decision to be made about if we want to have a button to use the search or only keyboard 'enter'

Ron-Lavi commented:
Any update design-wise on this prototype? @MariSvirik thanks 🙏🏼

MariSvirik commented:
@laviro Not really.
PF4 is working on their own design ideations/research - cutting this complex search input monster into smaller chunks such as bookmarks/ supporting value-attribute pairs/ etc. But these parts aren't stable designs neither they are compatible with each other yet. 
For searching action in a search input, PF4 currently uses [an arrow icon](https://www.patternfly.org/v4/components/search-input/design-guidelines/#advanced-search) - which may not be super intuitive when presented with other icons as well.

Here are a couple of options with a bookmark button and a search button as well.
<img width="369" alt="Screenshot 2021-09-20 at 11 08 34" src="https://user-images.githubusercontent.com/51960547/133979198-e853b3db-ad4f-4545-84d7-f8b8988c53df.png">

Option 1/2 is more consistent with current PF4 search input designs. Option 4 is kind of what we currently use in some places in Satellite.
I would rather not use a search button at all - it's cleaner, less confusing.

MariaAga commented:
@MariSvirik I think it's more confusing not having a search button, as without it the search is just a random text box and it won't search until the user will click enter.
Also option 1,2 button suggest that the page will go somewhere instead of search something which is not the case in foreman.

Ron-Lavi commented:
I think we can add the button for now and just upgrade the look to PF4 so it won't block other parts of the page to be upgraded, e.g: the design of the current searchbar doesn't look good with new  PF4 buttons aside.

tbrisker commented on webpack/assets/javascripts/react_app/routes/Audits/AuditsPage/AuditsPage.js:
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { Button, Icon } from 'patternfly-react';
+import { Button } from '@patternfly/react-core';
```
This change doesn't seem to be related to the searchbar styling. I get that you want to make the buttons look similar, but that can be done in a different PR.

tbrisker commented:
Regarding the search button - I agree, option 4 looks better. I don't think removing it makes sense, users are used to having a button they can click. It makes sense to not have a button when a search field is used for a fast (client-side) filter that applies as they type, but when they need to press enter to search it makes sense to have the button at least for now. 

Also, test failures are related.

Ron-Lavi commented:
This is how the search-bar looks now with PF4 styles:
![pf4_search](https://user-images.githubusercontent.com/26363699/136371484-0c95974f-442c-4a23-9fa2-79370af3edd4.png)

and the current one for older pages:
![Screenshot (29)](https://user-images.githubusercontent.com/26363699/136371545-fb3348bd-b344-4e7e-9170-f3c7016aca96.png)

MariSvirik commented:
Can we just get rid of everything that's inside the search field?
/, Filter..., x and just write Search there?

It would be an easy fix to change 'Filter...' to 'Search'
Plus X button doesn't make sense if the search bar is empty. It can't do anything. It should appear if there is an input...

MariaAga commented:
@laviro to fit the new design should this pr change the icon of the bookmarks?
Also, the new bookmark component is not complete as there were features added to the old one after the new was merged:
https://projects.theforeman.org/issues/32656

Ron-Lavi commented:
> Can we just get rid of everything that's inside the search field? /, Filter..., x and just write Search there?
> 
> It would be an easy fix to change 'Filter...' to 'Search' Plus X button doesn't make sense if the search bar is empty. It can't do anything. It should appear if there is an input...

@MariSvirik I can try to display the `X` button only when there's some query,
and we can also rename the input's placeholder,
but as for removing the slash shortcut... it might need more feedback and research in the community - and I prefer we will do it properly when writing the component from scratch.

> @laviro to fit the new design should this pr change the icon of the bookmarks?
> Also, the new bookmark component is not complete as there were features added to the old one after the new was merged:
> https://projects.theforeman.org/issues/32656

@MariaAga if the PF4 bookmark isn't fully ready yet, I tend to lean towards just replacing the styles for the old PF3 bookmarks component until the new one is ready.

@tbrisker, let's make the searchbar consistent, I think it will look a bit weird at the start on old pages - but it might give us a good push towards PF4 and full react pages.

MariaAga commented:
I agree about just changing the css for. Any way, opened https://github.com/theforeman/foreman/pull/8824 to fix icon+ add mange bookmarks.

MariSvirik commented:
@laviro sounds great. I can live with slash :D 
After discussion with our UX group, they opted for the approach to have a new search also on the legacy pages, which would pressure you to update the rest of the pages pretty soon.

Ron-Lavi commented:
> Any way, opened #8824 to fix icon+ add mange bookmarks.

NIce @MariaAga, that was fast!
actually, I tried manipulating the old bookmark to look like PF4 one, but it took too much work and too many file changes
so I used the commit from #8824 and it works great!

this is how the search bar looks at the moment:
![Screenshot (32)](https://user-images.githubusercontent.com/26363699/136702495-e464f823-093a-42f0-bcd9-4dc5821d4472.png)

Also, I wanted to remove the old bookmarks folder, though it seems the new PF4 bookmarks is trying to import components from there, so let's do it as a follow up and keep this PR small.

any other thoughts or concerns?

tbrisker commented:
@laviro please rebase, the other PR has been merged.

Ron-Lavi commented:
Rebased, thanks @MariaAga @tbrisker !
@MariSvirik I also removed the dots and the italic style to the placeholder:
![Screenshot (33)](https://user-images.githubusercontent.com/26363699/136785325-13bbb22f-8c9e-422f-88c8-1a92d4e73f14.png)

MariaAga commented:
@laviro  Great to see that the bookmarks work with the "new" search!
The new search looks short, and it's not taking it's full length of 500px even, can it be longer?
![Screenshot from 2021-10-11 14-07-26](https://user-images.githubusercontent.com/30431079/136787410-0ff0e086-8a91-40df-9b38-a4811f64ae64.png)
![Screenshot from 2021-10-11 14-06-02](https://user-images.githubusercontent.com/30431079/136787523-d45e817e-bcb9-4839-baa2-9c9c04a6288a.png)

Ron-Lavi commented:
Thanks @MariaAga ! makes sense, I will try to make it larger

tbrisker commented:
@laviro test failures are related

Ron-Lavi commented:
@MariaAga I just fixed the search bar width:

![gnome-shell-screenshot-99Z0A1](https://user-images.githubusercontent.com/26363699/136832566-5e69d0cd-63cc-417f-a461-8e3bab8f6e38.png)

MariaAga commented on webpack/assets/javascripts/react_app/components/PF4/Bookmarks/Bookmarks.js:
```diff
@@ -12,6 +12,7 @@ import {
 } from './BookmarkItems';
 import history from '../../../history';
 import { stringifyParams } from '../../../common/urlHelpers';
+import { translate as __ } from '../../../common/I18n';
```
Did the bookmarks title got deleted? I think it was a good idea to specify that " this icon is a button to bookmarks"

Ron-Lavi replied:
probably during the rebase

Ron-Lavi replied:
added it again 👍🏼

ronlavi2412 committed:
Fixes #32750 - make searchbar css looks like PF4

Ron-Lavi commented:
ready for another review round :)

MariaAga reviewed (approved):
Looka good to me :)
@tbrisker any other concerns?

tbrisker reviewed (approved):
Thanks @laviro and @MariaAga !
